pkg/stanza/fileconsumer/file_threadpool.go

```diff
@@ -0,0 +1,201 @@
+	m.syncLastPollFilesConcurrent(ctx)
+}
+
+func (m *Manager) worker(ctx context.Context) {
+	defer m.workerWg.Done()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case wrapper, ok := <-m.readerChan:
+			if !ok {
+				return
+			}
+			r, fp := wrapper.reader, wrapper.trieKey
+			if !m.readToEnd(ctx, r) && !wrapper.close {
+				// Save off any files that were not fully read or if deleteAfterRead is disabled.
+				m.knownFilesLock.Lock()
+				m.knownFiles = append(m.knownFiles, r)
+				m.knownFilesLock.Unlock()
+			} else if wrapper.close {
+				// this is a lost reader, close it and release the file descriptor
+				r.Close()
+			}
+			m.updateTrie(fp, false)
+		}
+	}
+}
+
+func (m *Manager) makeReaderConcurrent(filePath string) (*reader, *fingerprint.Fingerprint) {
+	fp, file := m.makeFingerprint(filePath)
+	if fp == nil {
+		return nil, nil
+	}
+
+	// check if the current file is already being consumed
+	if m.isCurrentlyConsuming(fp) {
+		if err := file.Close(); err != nil {
+			m.Errorf("problem closing file", "file", file.Name())
+		}
+		return nil, nil
+	}
+
+	// Exclude any empty fingerprints or duplicate fingerprints to avoid doubling up on copy-truncate files
+	if m.checkDuplicates(fp) {
+		if err := file.Close(); err != nil {
+			m.Errorf("problem closing file", "file", file.Name())
+		}
+		return nil, nil
+	}
+	m.currentFps = append(m.currentFps, fp)
+
+	reader, err := m.newReaderConcurrent(file, fp)
+	if err != nil {
+		m.Errorw("Failed to create reader", zap.Error(err))
+		return nil, nil
+	}
+	return reader, fp
+}
+
+func (m *Manager) consumeConcurrent(ctx context.Context, paths []string) {
+	m.clearOldReadersConcurrent(ctx)
+	for _, path := range paths {
+		reader, fp := m.makeReaderConcurrent(path)
+		if reader != nil {
+			// add fingerprint to trie
+			m.updateTrie(fp, true)
+			m.readerChan <- readerWrapper{reader: reader, trieKey: fp}
+		}
+	}
+}
+
+func (m *Manager) isCurrentlyConsuming(fp *fingerprint.Fingerprint) bool {
+	m.trieLock.RLock()
+	defer m.trieLock.RUnlock()
+	return m.trie.HasKey(fp.FirstBytes)
+}
+
+func (m *Manager) updateTrie(fp *fingerprint.Fingerprint, insert bool) {
+	m.trieLock.Lock()
+	defer m.trieLock.Unlock()
+	if insert {
+		m.trie.Put(fp.FirstBytes)
+	} else {
+		m.trie.Delete(fp.FirstBytes)
+	}
+}
+
+func (m *Manager) clearOldReadersConcurrent(ctx context.Context) {
+	m.knownFilesLock.Lock()
+	defer m.knownFilesLock.Unlock()
+	// Clear out old readers. They are sorted such that they are oldest first,
+	// so we can just find the first reader whose poll cycle is less than our
+	// limit i.e. last 3 cycles, and keep every reader after that
+	oldReaders := make([]*reader, 0)
+	i := 0
+	for ; i < len(m.knownFiles); i++ {
+		reader := m.knownFiles[i]
+		if reader.generation >= 3 {
+			oldReaders = append(oldReaders, reader)
+		} else {
+			break
+		}
+	}
+	m.knownFiles = m.knownFiles[i:]
+
+	for _, r := range oldReaders {
+		if m.isCurrentlyConsuming(r.Fingerprint) {
+			r.Close()
+		} else {
```

Comment:
> What if we are not currently reading the file, but we did read the file again since this reader was created?

I don't think so if that's possible. Because `m.clearOldReadersConcurrent(ctx)` updates the `knownFiles` and clears out old readers ***before*** the readers for current poll cycles are created.

The current `poll()` will begin **only** when the previous poll cycle has submitted all the readers to the thread pool and the trie will be up to date.
So, there's no chance that we'd add any new file to read between the previous `poll()` and current `clearOldReadersConcurrent()`.

Reply:
> I don't think so if that's possible. Because m.clearOldReadersConcurrent(ctx) updates the knownFiles and clears out old readers before the readers for current poll cycles are created.

But it only removes readers from 3 poll cycles ago, so readers from the previous cycle are still there.

Reply:
Yeah, I agree. `clearOldReaders`  won't touch those readers which are from previous cycles i.e. less than 3 cycles.

Reply:
and we also have a trie check in place, we won't accidentally push the same reader again.

Reply:
one more thing to note,
If a reader is in `knownFiles` array, it would effectively mean that it isn't being consumed at the moment.

Reply:
> > I don't think so if that's possible. Because m.clearOldReadersConcurrent(ctx) updates the knownFiles and clears out old readers before the readers for current poll cycles are created.
> 
> But it only removes readers from 3 poll cycles ago, so readers from the previous cycle are still there.

I still think this is a problem. When we `ReadToEnd` as the file is discarded, it will start from an out of date offset. This will cause duplicate ingestion of data.

Reply:
Dan, if you don't mind, can you give me a walk through with an example? I think we'll be able to sort it out fairly quickly then.

Reply:
Sure, I'll try to be concise so hopefully this is clear enough.

Let's make the situation unrealistically simple and say that one line is written to a file in between each poll cycle. So at each time, we create a reader that reads to the end of the file. Each of these is a reader for the same file, but each is in a different generate and _has a different offset_.

| Time | Actual Content | Offset After Reading |
| --- | --- | --- |
| t0 | `foo0\n` | 5 |
| t1 | `foo0\nfoo1\n` | 10 |
| t2 | `foo0\nfoo1\nfoo2\n` | 15 |
| t3 | `foo0\nfoo1\nfoo2\nfoo3\n` | 20 |

So at `t3`, when we decide to discard the reader from `t0`, if we `ReadToEnd` _on that reader_, we will read `foo1\n`, `foo2\n...` again, even though `foo1\n` was already read at `t1`, `foo2\n` was read at `t2`, etc.

Reply:
I see. What I can do is, I will create a test case which simulates this and check it out. Should be simple.
Thanks 😊

Reply:
With a tweak in logic and respecting the `max_concurrent_files`, we don't hold readers open for 3 `poll()` cycles anymore. So, I think we're good now.

To discuss the above-mentioned scenario, we remove a reader from `m.knownFiles` if we detect a match [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/stanza/fileconsumer/file.go#L301). 
Also, while copying a reader, we don't copy the `generation`. 
So, every time a reader is ***copied***, we remove the old reader from `m.knownFiles` and create a fresh reader with `generation` set to `0`.

The scenario would look something like this with knownFiles:

Time | Actual Content | Offset After Reading | knownFiles (before poll()) | knownFiles (after poll())
-- | -- | -- | -- | --
t0 | foo0\n | 5 | [] | [reader{offset: 5, generation: 0}]
t1 | foo0\nfoo1\n | 10 | [reader{offset: 5, generation: 0}] | [reader{offset: 10 generation: 0}]
t2 | foo0\nfoo1\nfoo2\n | 15 |  [reader{offset: 15, generation: 0}] | [reader{offset: 15, generation: 0}]
t3 | foo0\nfoo1\nfoo2\nfoo3\n | 20 |  [reader{offset: 20, generation: 0}] | [reader{offset: 20 generation: 0}]

Note: It is assumed that we finish reading asynchronously before the next poll() is executed.